File: server/ops/diagnostics/checks/capture-health.js
// Is the Self-Model's own capability recorder (self/self-capture.js's
// recordAttempt()) actually working? Reuses self-store.js's
// captureHealthSummary() directly rather than re-deriving the same
// capture_health rollup a second time — that table already exists
// specifically to tell apart "a broken recorder" from "a tool genuinely
// never used" (see root CLAUDE.md's Self-Model section). A HIGH FAILURE
// RATE here (not just >0 failures — a single transient failure is normal)
// is the honest signal that something is systemically wrong with the
// recorder itself, not with whatever it's recording.

import { captureHealthSummary } from '../../../self/self-store.js';

export const id = 'capture-health';
const MIN_ATTEMPTS_FOR_SIGNAL = 5; // below this, a rate is meaningless noise
const FAILURE_RATE_THRESHOLD = 0.5; // a MAJORITY failing is systemic, not incidental

export async function probe() {
  const summary = captureHealthSummary();
  if (summary.attempts24h < MIN_ATTEMPTS_FOR_SIGNAL) return { ok: true }; // not enough real data to say anything honest

  const rate = summary.failures24h / summary.attempts24h;
  if (rate > FAILURE_RATE_THRESHOLD) {
    return {
      ok: false,
      detail: `The Self-Model's own capability recorder has failed ${summary.failures24h} of ${summary.attempts24h} attempts in the last 24h (last failure: ${summary.lastFailureSource || 'unknown source'} at ${summary.lastFailureAt || 'unknown time'}).`,
    };
  }
  return { ok: true };
}

// No remedy() — a systemically failing recorder needs the owner to look at
// what changed (a schema issue, a disk problem), not an automatic retry.

File: backend/jarvis/ops/diagnostics/checks/capture_health.py
"""Is the thing that records what capabilities did still working?

A broken recorder and a capability genuinely never used look identical from
every dimension that reads the stats — which is why the recorder logs its own
successes and failures, and why this reads that log rather than the stats.
"""

from __future__ import annotations

from typing import Any

from ..registry import Check

#: Any failure at all is worth reporting: this records a handful of rows a turn,
#: so a failure is a real fault rather than an unlucky one.
FAILURE_THRESHOLD = 1


def probe() -> dict[str, Any]:
    from ....self import store as self_store

    summary = self_store.capture_health_summary(hours=24)
    failures = int(summary.get("failures") or 0)
    if failures >= FAILURE_THRESHOLD:
        return {"ok": False,
                "detail": (f"The capability recorder failed {failures} time(s) in the last "
                           f"24 hours, so what it says about reliability is incomplete.")}
    return {"ok": True}


CHECK = Check(id="capture_health", probe=probe)
